pkg/controllers/management/aks/aks_cluster_handler.go

```diff
@@ -0,0 +1,426 @@
+			return e.SetUnknown(cluster, apimgmtv3.ClusterConditionUpdated, "")
+		case <-timeout.C:
+			cluster, err = e.recordAppliedSpec(cluster)
+			if err != nil {
+				return cluster, err
+			}
+			return cluster, nil
+		}
+	}
+}
+
+// generateAndSetServiceAccount uses the API endpoint and CA cert to generate a service account token. The token is then copied to the cluster status.
+func (e *aksOperatorController) generateAndSetServiceAccount(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	restConfig, err := e.getKubeConfig(cluster)
+	if err != nil {
+		return cluster, fmt.Errorf("error getting service account token: %v", err)
+	}
+
+	clusterDialer, err := e.ClientDialer.ClusterDialer(cluster.Name)
+	if err != nil {
+		return cluster, err
+	}
+
+	saToken, err := generateSAToken(restConfig, clusterDialer)
+	if err != nil {
+		return cluster, fmt.Errorf("error getting service account token: %v", err)
+	}
+
+	cluster = cluster.DeepCopy()
+	cluster.Status.ServiceAccountToken = saToken
+	return e.ClusterClient.Update(cluster)
+}
+
+// buildAKSCCCreateObject returns an object that can be used with the kubernetes dynamic client to
+// create an AKSClusterConfig that matches the spec contained in the cluster's AKSConfig.
+func buildAKSCCCreateObject(cluster *mgmtv3.Cluster) (*unstructured.Unstructured, error) {
+	aksClusterConfig := aksv1.AKSClusterConfig{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "AKSClusterConfig",
+			APIVersion: aksV1,
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: cluster.Name,
+			OwnerReferences: []v1.OwnerReference{
+				{
+					Kind:       cluster.Kind,
+					APIVersion: rbac.RancherManagementAPIVersion,
+					Name:       cluster.Name,
+					UID:        cluster.UID,
+				},
+			},
+		},
+		Spec: *cluster.Spec.AKSConfig,
+	}
+
+	// convert AKS cluster config into unstructured object so it can be used with dynamic client
+	aksClusterConfigMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&aksClusterConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &unstructured.Unstructured{
+		Object: aksClusterConfigMap,
+	}, nil
+}
+
+// recordAppliedSpec sets the cluster's current spec as its appliedSpec
+func (e *aksOperatorController) recordAppliedSpec(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	if reflect.DeepEqual(cluster.Status.AppliedSpec.AKSConfig, cluster.Spec.AKSConfig) {
+		return cluster, nil
+	}
+
+	cluster = cluster.DeepCopy()
+	cluster.Status.AppliedSpec.AKSConfig = cluster.Spec.AKSConfig
+	return e.ClusterClient.Update(cluster)
+}
+
+// deployAKSOperator looks for the rancher-aks-operator app in the cattle-system namespace, if not found it is deployed.
+// If it is found but is outdated, the latest version is installed.
+func (e *aksOperatorController) deployAKSOperator() error {
+	return e.DeployOperator(aksOperator, aksOperatorTemplate, aksShortName)
+}
+
+func (e *aksOperatorController) generateSATokenWithPublicAPI(cluster *mgmtv3.Cluster) (string, bool, error) {
```

Comment:
The return type here needs to change to indicate that the function was not able to determine if Rancher can contact the cluster.
```suggestion
func (e *aksOperatorController) generateSATokenWithPublicAPI(cluster *mgmtv3.Cluster) (string, *bool, error) {
```